sound/soc/sof/intel/hda-dsp.c

```diff
@@ -412,6 +412,24 @@ static int hda_dsp_update_d0i3c_register(struct snd_sof_dev *sdev, u8 value)
 	return 0;
 }
 
+static bool hda_dsp_d0i3_playback_active(struct snd_sof_dev *sdev)
+{
+	struct snd_pcm_substream *substream;
+	struct snd_sof_pcm *spcm;
+	int dir;
+
+	list_for_each_entry(spcm, &sdev->pcm_list, list) {
+		substream = spcm->stream[SNDRV_PCM_STREAM_PLAYBACK].substream;
+		if (!substream || !substream->runtime)
+			continue;
+
+		if (spcm->stream[dir].d0i3_compatible)
+			return true;
```

Comment:
the logic seems flawed. If there is ONE stream that isn't d0i3_compatible then we can't use d0i3. You select d0i3 on the first compatible stream found. That isn't quite right?

Reply:
@plbossart  .  If we get here, all the stream are d0i3_compatible confirmed by snd_sof_dsp_only_d0i3_compatible_stream_active, but maybe some of them are capture.  So if one of them is playback, we need to set streaming flag.

Reply:
I will add comments

Reply:
dir is never initialized in this function. WHat is its value?

Reply:
I insist that the definition of D0i3 compatible is invalid.

We are confusing D0i3-hotwording, which works WITHOUT any DMA transfers while in D0i3, and D0i3-streaming which requires DMA transfers to work - albeit in bursts.

Conflating D0i3-hotwording and D0i3-capture is not correct. In the former the DMAs are not enabled in the latter they are.

Reply:
@RanderWang I'd reverse the check inside the loop. Basically what you want to know is whether ONLY D0i3 compatible stream is active and not aything else. so what you need is this:
```
/* check if only D0i3 playback stream is active. */
static bool hda_dsp_only_d0i3_playback_active(struct snd_sof_dev *sdev)
{
	struct snd_pcm_substream *substream;
	struct snd_sof_pcm *spcm;
	int dir;

	list_for_each_entry(spcm, &sdev->pcm_list, list) {
		for_each_pcm_streams(dir) {
			substream = spcm->stream[dir].substream;
			if (!substream || !substream->runtime)
				continue;

			if (!spcm->stream[dir].d0i3_compatible || dir == SNDRV_PCM_STREAM_CAPTURE)
				return false;
		}
	}

	return true;
}
```

Reply:
the code is out data. discard this thread

Reply:
> I insist that the definition of D0i3 compatible is invalid.
> 
> We are confusing D0i3-hotwording, which works WITHOUT any DMA transfers while in D0i3, and D0i3-streaming which requires DMA transfers to work - albeit in bursts.
> 
> Conflating D0i3-hotwording and D0i3-capture is not correct. In the former the DMAs are not enabled in the latter they are.

@plbossart  I fully agree with your idea about two cases.  Do you recommend to implement a fully complete function for D0i3-streaming detection like the one suggested by Ranjani ?   BTW if D0i3-hotwording and D0i3-streaming are both enabled, we need to set power gate and streaming flag in one setD0i3 ipc msg.

Reply:
refine the above example
````c
/* check if only D0i3 playback stream is active. */
static bool hda_dsp_only_d0i3_playback_active(struct snd_sof_dev *sdev)
{
	struct snd_pcm_substream *substream;
	struct snd_sof_pcm *spcm;
        bool playback_active = false;
	int dir;

	list_for_each_entry(spcm, &sdev->pcm_list, list) {
		for_each_pcm_streams(dir) {
			substream = spcm->stream[dir].substream;
			if (!substream || !substream->runtime)
				continue;

			if (!spcm->stream[dir].d0i3_compatible)
				return false;

                        if (dir == SNDRV_PCM_STREAM_PLAYBACK)
                            playback_active = true;
		}
	}

	return playback_active ;
}
````

Reply:
@RanderWang i guess the question is are the flags the same for playback active and capture active when entering D0i3? and can they be used simultaneously? If not, this logic is incorrect

Reply:
@ranj063 thanks! D0i3-streaming (playback) would enable host dma and this doesn't affect D0i3-capture.  All flags are defined in
````c
		struct {
			/* legacy wake type (see WakeType) */
			uint32_t wake			: 1;
			/* streaming active now */
			uint32_t streaming		: 1;
			/* D0/D0ix transitions allowed (PG disabled) */
			uint32_t prevent_power_gating : 1;
			/* Clock gating enabled */
			uint32_t prevent_local_clock_gating : 1;

			uint32_t rsvd1			: 26;
			uint32_t _reserved_2	: 2;
		} r;
````

Reply:
@plbossart  D0i3-hotwording and D0i3-streaming  share the d0i3 setting in host driver and power gate setting in audio hardware except host dma. D0i3compatible means driver can go d0i3 path but need to check more condition like streaming.

Now we set d0i3compatible flag in topology and check it. If all the active streams are set this flag, we will check whether there is a playback and enable streaming flag.  Did you mean we need to introduce another flag D0i3streaming or something else like above discussing.   Thanks!